### PR TITLE
Batch nonce for channels

### PR DESCRIPTION
**Pull request by vovac12**

### CODE REVIEW

File: pallets/types/src/evm.rs [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    Eq,
+    scale_info::TypeInfo,
+    codec::MaxEncodedLen,
+)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+/// Information about asset in EVM network
+pub struct EVMAssetInfo {
+    /// Thischain asset id
+    pub asset_id: MainnetAssetId,
+    /// Contract address
```
Comment (Alexey-N-Chernyshov) [inline pallets/types/src/evm.rs:123]: Is it ERC-20 contract of asset?

File: pallets/types/Cargo.toml [hunk 2]
```diff
@@ -50,28 +51,8 @@ serde_json = "1.0"
...
-  "parity-bytes/std",
-  "rlp/std",
-  "sp-core/std",
-  "sp-io/std",
-  "sp-runtime/std",
-  "sp-std/std",
-  "sp-beefy/std",
-  "xcm/std",
-]
+std = ["serde", "codec/std", "scale-info/std", "ethabi/std", "ethbloom/std", "ethereum-types/std", "hex/std", "libsecp256k1/std", "parity-bytes/std", "rlp/std", "sp-core/std", "sp-io/std", "sp-runtime/std", "sp-std/std", "sp-beefy/std", "xcm/std"]
```
Comment (Alexey-N-Chernyshov) [inline pallets/types/Cargo.toml]: Previously was more readable.

File: pallets/substrate-channel/src/outbound/mod.rs [hunk 3]
```diff
@@ -186,65 +181,64 @@ pub mod pallet {
     }
 
     impl<T: Config> Pallet<T> {
-        pub fn make_message_id(nonce: u64) -> H256 {
-            MessageId::outbound(nonce).using_encoded(|v| <T as Config>::Hashing::hash(v))
+        pub fn make_message_id(batch_nonce: u64, message_nonce: u64) -> H256 {
+            MessageId::outbound_batched(batch_nonce, message_nonce).hash()
```
Comment (Alexey-N-Chernyshov) [inline pallets/substrate-channel/src/outbound/mod.rs:185]: If non-batched message ids are not necessary anymore, it is better to remove `MessageId::outbound()` builder in `sora2-common`.

File: pallets/substrate-channel/src/outbound/mod.rs [hunk 4]
```diff
@@ -186,65 +181,64 @@ pub mod pallet {
...
+            let batch_nonce = ChannelNonces::<T>::mutate(network_id, |nonce| {
+                *nonce += 1;
+                *nonce
+            });
+
+            for idx in 0..messages.len() as u64 {
                 T::MessageStatusNotifier::update_status(
                     GenericNetworkId::Sub(network_id),
-                    Self::make_message_id(message.nonce),
+                    Self::make_message_id(batch_nonce, idx as u64),
```
Comment (Alexey-N-Chernyshov) [inline pallets/substrate-channel/src/outbound/mod.rs]:
```suggestion
                    Self::make_message_id(batch_nonce, idx),
```
`idx` is u64

File: pallets/substrate-channel/src/outbound/mod.rs [hunk 5]
```diff
@@ -277,34 +271,37 @@ pub mod pallet {
             _: (),
         ) -> Result<H256, DispatchError> {
             debug!("Send message from {:?} to network {:?}", who, network_id);
+            let messages_count = MessageQueues::<T>::decode_len(network_id).unwrap_or(0);
```
Comment (Alexey-N-Chernyshov) [inline pallets/substrate-channel/src/outbound/mod.rs]:
```suggestion
            let messages_count: u64 = MessageQueues::<T>::decode_len(network_id).unwrap_or(0);
```

File: pallets/substrate-channel/src/outbound/mod.rs [hunk 6]
```diff
@@ -277,34 +271,37 @@ pub mod pallet {
...
             ensure!(
-                MessageQueues::<T>::decode_len(network_id).unwrap_or(0)
-                    < T::MaxMessagesPerCommit::get() as usize,
+                messages_count < T::MaxMessagesPerCommit::get() as usize,
```
Comment (Alexey-N-Chernyshov) [inline pallets/substrate-channel/src/outbound/mod.rs]:
```suggestion
                messages_count < T::MaxMessagesPerCommit::get() as u64,
```

File: pallets/substrate-channel/src/outbound/mod.rs [hunk 7]
```diff
@@ -277,34 +271,37 @@ pub mod pallet {
...
+                        .try_into()
+                        .map_err(|_| Error::<T>::PayloadTooLarge)?,
+                    timepoint: T::TimepointProvider::get_timepoint(),
+                },
+            )
+            .map_err(|_| Error::<T>::QueueSizeLimitReached)?;
+            Self::deposit_event(Event::MessageAccepted {
+                network_id,
+                batch_nonce,
+                message_nonce: messages_count as u64,
```
Comment (Alexey-N-Chernyshov) [inline pallets/substrate-channel/src/outbound/mod.rs]:
```suggestion
                message_nonce: messages_count,
```

File: pallets/substrate-channel/src/outbound/mod.rs [hunk 8]
```diff
@@ -277,34 +271,37 @@ pub mod pallet {
...
+            });
+            Ok(Self::make_message_id(batch_nonce, messages_count as u64))
```
Comment (Alexey-N-Chernyshov) [inline pallets/substrate-channel/src/outbound/mod.rs]:
```suggestion
            Ok(Self::make_message_id(batch_nonce, messages_count))
```

File: pallets/substrate-channel/src/inbound/mod.rs [hunk 9]
```diff
@@ -125,28 +128,31 @@ pub mod pallet {
...
-                        *nonce += 1;
-                        Ok(())
-                    }
-                })?;
-
-                // Self::handle_fee(message.fee.unique_saturated_into(), &relayer);
-
-                let message_id = MessageId::inbound(message.nonce);
+            T::Verifier::verify(network_id.into(), commitment_hash, &proof)?;
+            // Verify message nonce
```
Comment (Alexey-N-Chernyshov) [inline pallets/substrate-channel/src/inbound/mod.rs]:
```suggestion
            // Verify batch nonce
```

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Review by N1ghtStorm — APPROVED** [review]

File: pallets/types/src/evm.rs (hunk 1 above)
Comment (vovac12) [reply pallets/types/src/evm.rs:123]: Yes. Added info to comment

**Pull request merged**